core/state_processor.go

```diff
@@ -170,14 +194,22 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 						return nil, nil, 0, fmt.Errorf("error decoding bytes %x: %w", stIt.Slot(), err)
 					}
 					copy(safeValue[32-len(value):], value)
-					slotnr := rawdb.ReadPreimage(statedb.Database().DiskDB(), stIt.Hash())
 
-					mkv.addStorageSlot(addr, slotnr, safeValue[:])
+					var slotnr [32]byte
+					if _, err := io.ReadFull(fpreimages, slotnr[:]); err != nil {
+						return nil, nil, 0, fmt.Errorf("reading preimage file: %s", err)
+					}
+					if crypto.Keccak256Hash(slotnr[:]) != stIt.Hash() {
+						return nil, nil, 0, fmt.Errorf("preimage file does not match storage hash: %s!=%s", crypto.Keccak256Hash(slotnr[:]), stIt.Hash())
+					}
```

Comment:
We double check that the preimage we received from the file, matches what we were looking for.

Strictly speaking, this shouldn't be needed if we trust the file is correct but honestly considering `Keccack256Hash` is very fast I prefer to be conservative.

We can later add some flag to make it an optional check.